README
= Super Exception Notifier

The Super Exception Notifier (SEN) gem provides a mailer object and a default set of templates 
for sending email notifications when errors occur in a Rails application, as well as a default 
set of error page templates to render based on the status code assigned to an error.  The gem 
is configurable, allowing programmers to customize (settings are per environment or per class):

* the sender address of the email
* the recipient addresses
* text used to prepend and append the subject line
* the HTTP status codes to send emails for
* the error classes to send emails for
* alternatively, the error classes to not send emails for
* whether to send error emails or just render without sending anything
* the HTTP status and status code that gets rendered with specific errors
* the view path to the error page templates
* custom errors, with custom error templates
* fine-grained customization of error layouts (or no layout)
* get error notification for errors that occur in the console, using notifiable method
* Hooks into `git blame` output so you can get an idea of who (may) have introduced the bug
* Hooks into other website services (e.g. you can send exceptions to to Switchub.com)
* Specify which level of notification you would like with an array of optional styles of notification:
 [:render, :email, :web_hooks]
* Can notify of errors occurring in any method in any class in Ruby by wrapping the method call like this:
 notifiable { method }
* Can notify of errors in Rake tasks using 'NotifiedTask.new' instead of 'task' when writing tasks
* Works with Hoptoad Notifier, so you can notify via SEN and/or Hoptoad for any particular errors.
* Tested with Rails 2.3.x, should work with Rails 2.2.x, and is apparently not yet compatible with Rails 3.

The email includes information about the current request, session, and
environment, and also gives a backtrace of the exception.

This gem is based on the wonderful exception_notification plugin created by Jamis Buck. I have 
modified it extensively and merged many of the improvements from a dozen or so other forks.  
It remains a (mostly) drop in replacement with greatly extended functionality and customization 
options.  I keep it up to date with the work on the core team's
branch.

The venerable original is here:
 
  http://github.com/rails/exception_notification

The current version of this gem is a git fork of the original and has been updated to include 
the latest improvements from the original, and many improvements from the other forks on github.  
I merge them in when I have time, and when the changes fit nicely with the enhancements I have 
already made.

This fork of Exception Notifier is in production use on several large websites (top 5000).

== Installation as RubyGem

  [sudo] gem install super_exception_notifier

More Installation Options are here: http://wiki.github.com/pboling/exception_notification/installation

== Configuration as RubyGem

  config.gem 'super_exception_notifier', :lib => "exception_notification"

More Configuration Options are here: http://wiki.github.com/pboling/exception_notification/configuration

== Configuration In Environment (Initializer)

  ExceptionNotification::Notifier.configure_exception_notifier do |config|
    config[:app_name]                 = "[MYAPP]"
    config[:sender_address]           = "super.exception.notifier@example.com"
    config[:exception_recipients]     = [] # You need to set at least one recipient if you want to get the notifications
    # In a local environment only use this gem to render, never email
    #defaults to false - meaning by default it sends email.  Setting true will cause it to only render the error pages, and NOT email.
    config[:skip_local_notification]  = true
    # Error Notification will be sent if the HTTP response code for the error matches one of the following error codes
    config[:notify_error_codes]   = %W( 405 500 503 )
    # Error Notification will be sent if the error class matches one of the following error classes
    config[:notify_error_classes] = %W( )
    # What should we do for errors not listed?
    config[:notify_other_errors]  = true
    # If you set this SEN will attempt to use git blame to discover the person who made the last change to the problem code
    config[:git_repo_path]            = nil # ssh://git@blah.example.com/repo/webapp.git
  end

More Configuration Options: http://wiki.github.com/pboling/exception_notification/advanced-environment-configuration

== Handling Errors in Request Cycle

1. Include the ExceptionNotification::ExceptionNotifiable mixin in whichever controller you want to generate error emails (typically ApplicationController):

  class ApplicationController < ActionController::Base
    ############################################################
    # ERROR HANDLING et Foo
    include ExceptionNotification::ExceptionNotifiable
    #Comment out the line below if you want to see the normal rails errors in normal development.
    alias :rescue_action_locally :rescue_action_in_public if Rails.env == 'development'
    #self.error_layout = 'errors'
    self.exception_notifiable_verbose = true #SEN uses logger.info, so won't be verbose in production
    self.exception_notifiable_pass_through = :hoptoad # requires the standard hoptoad gem to be installed, and setup normally
    self.exception_notifiable_silent_exceptions = [Acl9::AccessDenied, MethodDisabled, ActionController::RoutingError ]
    #specific errors can be handled by something else:
    rescue_from 'Acl9::AccessDenied', :with => :access_denied
    # END ERROR HANDLING
    ############################################################
    ...
  end
  
2. Specify the email recipients in your environment (You should have already done this in the "Configuration in Environment (Initializer)" step above):

  ExceptionNotification::Notifier.configure_exception_notifier do |config|
    config[:exception_recipients] = %w(joe@example.com bill@example.com)
  end

3. Make sure you have your ActionMailer server settings correct if you are using the e-mail features.

4. That’s it! The defaults take care of the rest.

http://wiki.github.com/pboling/exception_notification/exceptions-inside-request-cycle

== Advanced Configuration

There is a lot more you can configure, and do:

http://wiki.github.com/pboling/exception_notification/


CHANGELOG.txt
2014-01-18 v3.1
- Lots of refactoring, mostly in the test suite by Peter Boling
- Now tested against Rails 2.3 and 2.2 by Peter Boling
- Still 2 failing tests on Rails 2.1 and 2.0

2014-01-18 v3.0.16
- Fixed all the URLs in the Readme.

2014-01-18 v3.0.15
- Conform to gem-release standard VERSION namespace (SuperExceptionNotifier::VERSION)

2014-01-18 v3.0.14
- Fixed Notifiable class by Peter Boling
- Fixed ExceptionNotifiable class by Bradford Folkens
- Modernized lots of bits and pieces by Peter Boling
- All tests now pass in Rails 2.3.18 and Ruby 1.8.7-p374 by Peter Boling
- Fixes typos
- Switched to Bundler; Removed Jeweler
- Readme makeover

2010-06-30 v3.0.13
- Fixed homepage in Rakefile, noted Rails 3 incompatibility in README.

2010-06-25 v3.0.12
- Improved clarity of readme.

2010-06-18 v3.0.11
- Added correct view to Rakefile to include with build. (Jeweler had missed it because it was the wrong filename)

2010-06-18 v3.0.10
- Added view to gemspec to include with build. (Not sure why Jeweler missed it)

2010-06-18 v3.0.9
- Fixed bad view path for mailer.

2010 - Version 3 refactoring
- Realigned structure of gem to match the standard rails branch
- Added Hoptoad Notification Integration (requires standard Hoptoad Gem)

2009-01-28 v2.0.8
- improve readme

2009-12-30 v2.0.7
- cite /rails/app/views/exception_notifiable/500.html in Rakefile (oops!)

2009-12-30 v2.0.6
- cite /rails/app/views/exception_notifiable/500.html in gemspec (oops!)

2009-12-30 v2.0.5
- use blank? instead of empty? so nils won't cause an error

2009-11-04
* Fixing Typos in Comments and Readme
* General Readme cleanup


lib/exception_notifier.rb
require 'pathname'

class ExceptionNotifier < ActionMailer::Base

  #andrewroth reported that @@config gets clobbered because rails loads this class twice when installed as a plugin, and adding the ||= fixed it.
  @@config ||= {
    # If left empty web hooks will not be engaged
    :web_hooks                => [],
    :app_name                 => "[MYAPP]",
    :version                  => "0.0.0",
    :sender_address           => "super.exception.notifier@example.com",
    :exception_recipients     => [],
    # Customize the subject line
    :subject_prepend          => "[#{(defined?(Rails) ? Rails.env : RAILS_ENV).capitalize} ERROR] ",
    :subject_append           => nil,
    # Include which sections of the exception email? 
    :sections                 => %w(request session environment backtrace),
    :skip_local_notification  => true,
    :view_path                => nil,
    #Error Notification will be sent if the HTTP response code for the error matches one of the following error codes
    :notify_error_codes   => %W( 405 500 503 ),
    #Error Notification will be sent if the error class matches one of the following error error classes
    :notify_error_classes => %W( ),
    :notify_other_errors  => true,
    :git_repo_path            => nil,
    :template_root            => "#{File.dirname(__FILE__)}/../views"
  }

  cattr_accessor :config

  def self.configure_exception_notifier(&block)
    yield @@config
  end
  
  self.template_root = config[:template_root]

  def self.reloadable?() false end

  # Returns an array of potential filenames to look for
  # eg. For the Exception Class - SuperExceptionNotifier::CustomExceptionClasses::MethodDisabled
  # the filename handles are:
  #   super_exception_notifier_custom_exception_classes_method_disabled
  #   method_disabled
  def self.exception_to_filenames(exception)
    filenames = []
    e = exception.to_s
    filenames << ExceptionNotifier.filenamify(e)

    last_colon = e.rindex(':')
    unless last_colon.nil?
      filenames << ExceptionNotifier.filenamify(e[(last_colon + 1)..(e.length - 1)])
    end
    filenames
  end

  # Converts Stringified Class Names to acceptable filename handles with underscores
  def self.filenamify(str)
    str.delete(':').gsub( /([A-Za-z])([A-Z])/, '\1' << '_' << '\2').downcase
  end

  # What is the path of the file we will render to the user based on a given status code?
  def self.get_view_path_for_status_code(status_cd, verbose = false)
    file_name = ExceptionNotifier.get_view_path(status_cd, verbose)
    #ExceptionNotifierHelper::COMPAT_MODE ? "#{File.dirname(__FILE__)}/../rails/app/views/exception_notifiable/500.html" : "500.html"
    file_name.nil? ? self.catch_all(verbose) : file_name
  end

#  def self.get_view_path_for_files(filenames = [])
#    filepaths = filenames.map do |file|
#      ExceptionNotifier.get_view_path(file)
#    end.compact
#    filepaths.empty? ? "#{File.dirname(__FILE__)}/../rails/app/views/exception_notifiable/500.html" : filepaths.first
#  end

  # What is the path of the file we will render to the user based on a given exception class?
  def self.get_view_path_for_class(exception, verbose = false)
    return self.catch_all if exception.nil?
    return self.catch_all unless exception.is_a?(StandardError) || exception.is_a?(Class) # For some reason exception.is_a?(Class) works in console, but not when running in mongrel (ALWAYS returns false)?!?!?
    filepaths = ExceptionNotifier.exception_to_filenames(exception).map do |file|
      ExceptionNotifier.get_view_path(file, verbose)
    end.compact
    filepaths.empty? ? self.catch_all(verbose) : filepaths.first
  end

  def self.catch_all(verbose = false)
    puts "[CATCH ALL INVOKED] #{File.dirname(__FILE__)}/../rails/app/views/exception_notifiable/500.html" if verbose
    "#{File.dirname(__FILE__)}/../rails/app/views/exception_notifiable/500.html"
  end

  # Check the usual suspects
  def self.get_view_path(file_name, verbose = false)
    if File.exist?("#{RAILS_ROOT}/public/#{file_name}.html")
      puts "[FOUND FILE] #{RAILS_ROOT}/public/#{file_name}.html" if verbose
      "#{RAILS_ROOT}/public/#{file_name}.html"
    elsif !config[:view_path].nil? && File.exist?("#{RAILS_ROOT}/#{config[:view_path]}/#{file_name}.html.erb")
      puts "[FOUND FILE] #{RAILS_ROOT}/#{config[:view_path]}/#{file_name}.html.erb" if verbose
      "#{RAILS_ROOT}/#{config[:view_path]}/#{file_name}.html.erb"
    elsif !config[:view_path].nil? && File.exist?("#{RAILS_ROOT}/#{config[:view_path]}/#{file_name}.html")
      puts "[FOUND FILE] #{RAILS_ROOT}/#{config[:view_path]}/#{file_name}.html" if verbose
      "#{RAILS_ROOT}/#{config[:view_path]}/#{file_name}.html"
    elsif File.exist?("#{File.dirname(__FILE__)}/../rails/app/views/exception_notifiable/#{file_name}.html.erb")
      puts "[FOUND FILE] #{File.dirname(__FILE__)}/../rails/app/views/exception_notifiable/#{file_name}.html.erb" if verbose
      "#{File.dirname(__FILE__)}/../rails/app/views/exception_notifiable/#{file_name}.html.erb"
    elsif File.exist?("#{File.dirname(__FILE__)}/../rails/app/views/exception_notifiable/#{file_name}.html")
      #ExceptionNotifierHelper::COMPAT_MODE ? "#{File.dirname(__FILE__)}/../rails/app/views/exception_notifiable/#{file_name}.html" : "#{status_cd}.html"
      puts "[FOUND FILE] #{File.dirname(__FILE__)}/../rails/app/views/exception_notifiable/#{file_name}.html" if verbose
      "#{File.dirname(__FILE__)}/../rails/app/views/exception_notifiable/#{file_name}.html"
    else
      nil
    end
  end

  def exception_notification(exception, class_name = nil, method_name = nil, request = nil, data={}, the_blamed=nil)
    body_hash = error_environment_data_hash(exception, class_name, method_name, request, data, the_blamed)
    #Prefer to have custom, potentially HTML email templates available
    #content_type  "text/plain"
    recipients    config[:exception_recipients]
    from          config[:sender_address]

    request.session.inspect unless request.nil? # Ensure session data is loaded (Rails 2.3 lazy-loading)
    
    subject       "#{config[:subject_prepend]}#{body_hash[:location]} (#{exception.class}) #{exception.message.inspect}#{config[:subject_append]}"
    body          body_hash
  end
  
  def background_exception_notification(exception, data = {}, the_blamed = nil)
    exception_notification(exception, nil, nil, data, the_blamed)
  end

  def rake_exception_notification(exception, task, data={})
    content_type "text/plain"

    subject    "#{email_prefix}#{task.name} (#{exception.class}) #{exception.message.inspect}"

    recipients exception_recipients
    from       sender_address

    body       data.merge({ :task => task,
                  :exception => exception,
                  :backtrace => sanitize_backtrace(exception.backtrace),
                  :rails_root => rails_root,
                  :data => data,
                  :sections => sections.reject{|s| %w(request session).include?(s) } 
               })
  end

  private

    def error_environment_data_hash(exception, class_name = nil, method_name = nil, request = nil, data={}, the_blamed=nil)
      data.merge!({
        :exception => exception,
        :backtrace => sanitize_backtrace(exception.backtrace),
        :rails_root => rails_root,
        :data => data,
        :the_blamed => the_blamed
      })

      data.merge!({:class_name => class_name}) if class_name
      data.merge!({:method_name => method_name}) if method_name
      if class_name && method_name
        data.merge!({:location => "#{class_name}##{method_name}"})
      else
        data.merge!({:location => sanitize_backtrace([exception.backtrace.first]).first})
      end
      if request
        data.merge!({:request => request})
        data.merge!({:host => (request.env['HTTP_X_REAL_IP'] || request.env["HTTP_X_FORWARDED_HOST"] || request.env["HTTP_HOST"])})
        data.merge!({:sections => config[:sections]})
      else
        # TODO: with refactoring in the view structure, the environment section could show useful ENV data even without a request?
        data.merge!({:sections => config[:sections] - %w(request session environment)})
      end
      return data
    end

    def sanitize_backtrace(trace)
      re = Regexp.new(/^#{Regexp.escape(rails_root)}/)
      trace.map { |line| Pathname.new(line.gsub(re, "[RAILS_ROOT]")).cleanpath.to_s }
    end

    def rails_root
      @rails_root ||= Pathname.new(RAILS_ROOT).cleanpath.to_s
    end

end
